PIG 16 - Model Priors API (#4381)
gammapy/gammapy: docs/development/pigs/pig-026.rst

.. include:: ../../references.txt

.. _pig-026:

********************************
PIG 26 - Model Priors API 
********************************

* Author: Noah Biederbeck, Katrin Streil
* Created: ``2023-03-13``

> * Accepted:

Reviewer: ```suggestion
* Accepted: ``2023-11-03``
```

> * Status: draft

Reviewer: ```suggestion
* Status: Accepted
```

                          dataset_name = dataset.name)
  tilt_prior =  GaussianPrior(mu = "0", sigma = "0.05")
  norm_prior =  GaussianPrior(mu = "1", sigma = "0.1")

  bkg_model.set_prior([bkg_model.parameters['tilt'],bkg_model.parameters['norm']], [tilt_prior, norm_prior])

A preliminary version was set up like this and tests on simulated datasets were successful. Note that this setup is quite simple and it can be developed more advanced based on the same principle. 

Case 2: Favoring postivive values for flux amplitudes
--------------

A step-like prior function can be used to favour positive values for physical properties like the flux amplitude. By setting a prior one avoids defining hard boundary conditions with the ``min`` attribute of the to-be-fitted parameter. 
The prior is set to ``value`` if the parameter value is between ``xmin`` and ``xmax`` and 0 if not.

.. code:: python 

   from gammapy.modeling.models import PowerLawSpectraModel, StepPrior

   pwl = PowerLawSpectraModel()
   prior = StepPrior(xmin = "-inf", xmax = "0",  value = "1")
   pwl.set_prior([pwl.parameters['amplitude']], [prior])
  

> Case 3: Support unfolding methods for spectral flux points `#4122`_
> --------------

Reviewer: ```suggestion
Case 3: Support unfolding methods for spectral flux points `#4122`_
----------------------------------------------------------
```

            
        def __call__(self, value):
            """Call evaluate method"""
            kwargs = {par.name: par.quantity for par in self.parameters}
            if isinstance(value, Parameter):
                return self.evaluate(value.quantity, **kwargs)     
             else:
                raise TypeError(f"Invalid type: {value}, {type(value)}")
            

The base ``PriorModel`` class inherits from the ``ModelBase`` class and allows to set (for now one) unit to which the to-be-evaluated parameter the prior is set on is converged. In addition, there can be a weight set. The parameters of the  ``PriorModel`` are ``PriorParameters`` and ``PriorParameter``. They inherit from the ``Parameters`` and ``Parameter`` classes, respectively, however, have a limited amount of attributes. 
It is assumed that the ``PriorParameters`` are set in the same unit as the ``Parameter`` they get evaluated on.  

Future additional properties of the ``PriorModel`` classes: 

- Write/read from/to a yaml file, ideally also when the corresponding model is written/read (see serialisation example below)
- Prior registry system
- Different prior subclasses depending on the use cases 

> Exemplary additional prior subclasses:
> ---------------------------------------
> .. code:: python

Reviewer: ```suggestion
Exemplary additional prior subclasses:
--------------------------------------

.. code:: python
```

  class GaussianPrior(PriorModel):

        """Gaussian Prior with mu and sigma.
        """
        tag = ["GaussianPrior"]
        _type = "prior"
        mu = PriorParameter(name="mu", value = 0, unit = '')
        sigma = PriorParameter(name="sigma", value = 1, unit = '')
        
        @staticmethod
        def evaluate(value, mu, sigma):
            return ((value - mu) / sigma) ** 2

.. code:: python

  class UniformPrior(PriorModel):

        """Uniform Prior
        """
        tag = ["UniformPrior"]
        uni = PriorParameter(name="uni", value = 0, min = 0, max = 10, unit = '' )
        
        @staticmethod
        def evaluate(value, uni):
            return uni

>             

Reviewer 2: A prior that can be quite useful is the LogUniform prior.
For example for the normalization of a source, giving a linear prior between 1e-14 and 1e-11 is not ideal. Having the LogUniform priors will become handy if at some point we perform Bayesian analyses.
It could be something like this : 

```
class LogUniformPrior(PriorModel):
         """LogUniform Prior
         """
         tag = ["LogUniformPrior"]
         uni = PriorParameter(name="loguni", value = 1, min = 0.1, max = 10, unit = '' )
```

A note on the fact that in the Uniform and LogUniform priors the `value` parameter itself is a not useful as we just want to define (min, max) boundaries.
Is there a reason we should have a `value` parameter in this case ?

Reviewer 3: Thanks for the comment! 
Yes, I think this kind of prior could be helpful and is very straightforward to implement. 
The reason for the ``value`` parameter was to gain more flexibility in the use cases. In some cases, one might the prior to be stronger than in others.

Reviewer: Does this prior need a range for the parameter , e.g.:
```
 min = PriorParameter(name="min", value = -np.inf, unit = '' )
 max = PriorParameter(name="min", value = np.inf, unit = '' )
```

The priors can be set on a ```Parameter``` as ``.prior`` and get evaluated within:

.. code:: python 

  class Parameter():

        _prior = None
    
        @property
        def prior(self):
            return self._prior
    
        @prior.setter
        def prior(self, value):
            self._prior = value
    
        def prior_stat_sum(self):
            if self.prior is not None:
                return self.prior.weight * self.prior(self)

The ``Parameters`` inherit the priors and evaluate the ``prior_stat_sum`` of all the ``Parameters.parameters``:

.. code:: python 

  class Parameters():
   
        @property
        def prior(self):
              return [par.prior for par in self]
          
          def prior_stat_sum(self):
              parameters_stat_sum = 0
              for par in self:
                  if par.prior is not None:
                      parameters_stat_sum += par.prior_stat_sum()
              return parameters_stat_sum

> During the Fit the ``stat_sum`` of the Datasets is evaluated where now the statistics due to the priors set on the parameters is added. 

Reviewer: ```suggestion
During the Fit the ``Datasets.stat_sum()`` function is evaluated.  Now the function has to compute the statistics due to the priors set on the parameters and add to the stat_sum. 
```

                stat = stat[self.mask.data]
            prior_stat_sum = self.models.parameters.prior_stat_sum()
            return np.sum(stat, dtype=np.float64) + prior_stat_sum

A simple method in the ``Models`` class will allow setting priors on multiple parameters simultaneously. 

.. code:: python 

  class Models(Models):
      
        def set_prior(self, parameters, priors):
            for parameter, prior in zip(parameters, priors):
                parameter.prior = prior

Serialisation 
------------

The priors are serialised and read and writeable to a yaml file. They are also saved if the associated ``Parameter`` or ``Parameters`` is transformed to a dictionary. For this, the prior subclasses are tagged. The following two example shows the resulting yaml file for a ``Parameter``.

.. code-block:: yaml

  {'name': 'testpar',
   'value': 0.1,
   'unit': '',
   'error': 0,
   'min': nan,
   'max': nan,
   'frozen': False,
   'interp': 'lin',
   'scale_method': 'scale10',
   'is_norm': False,
   'prior': {'tag': 'GaussianPrior',
            'parameters': [{'name': 'mu', 'value': 0},
                           {'name': 'sigma', 'value': 0.1}]}}

> Implementation Outline
> --------------

Reviewer: ```suggestion
Implementation Outline
-----------------------
```

The PIG implementation will be piece-wise within individual issues in the following order:

1. ``Prior`` base class and subclasses (See `#4620`_)
2. ``PriorParameter`` and ``PriorParameters`` (See `#4620`_)
3. ``.prior`` attribute in the ``Parameter``, ``Parameters`` and ``Model`` classes
4.  Evaluation of the prior in the ``Datasets`` class

Decision
========

> 

Reviewer: The PIG was discussed and reviewed among Gammapy developers. It has been reviewed by the CC and is now considered accepted.

> 

Reviewer: ```suggestion
The PIG was discussed and reviewed among Gammapy developers. It has been reviewed by the CC and is now considered accepted.
```

.. include:: ../../references.txt

.. _pig-026:

********************************
PIG 26 - Model Priors API 
********************************

* Author: Noah Biederbeck, Katrin Streil
* Created: ``2023-03-13``
* Accepted:
* Status: draft
* Discussion: `#4381`_

Abstract
========

This PIG is intended to introduce priors on models or on parameters that are evaluated during fitting.

Motivation
==========

Using priors on models or on parameters is the next step in full-fledged statistical analyses. 
A prior can incorporate the analysers' knowledge of the topic at hand or information about
the estimated IRFs systematics provided by the corresponding experiment, and yield more realistic and trustworthy results. 
The proposed formalism also includes the application of nuisance parameters and regularization.

Use cases
=========

In the past priors were a regularly requested feature or the solution to a problem.
See the following issues and PRs of Gammapy:

- Add sigma v estimator functionality for dark matter use case `#2075`_
- Background systematics as a nuisance parameter `#3955`_
- Support unfolding methods for spectral flux points `#4122`_
- Add PiecewiseNormSpatialModel `#4208`_

> 

Reviewer 4: This is also my main remaining comment. Adding the uses cases in code. E.g. for the unfolding it could look like:

```python3
import numpy as np
from astropy import units as u
from gammapy.modeling.model import PiecewiseNormSpectralModel, MultivariateGaussianPrior, PowerLawSpectralModel

n_points = 10

pwl = PowerLawSpectraModel()

energy = np.geomspace(1, 10, n_points) * u.TeV
norm =  PiecewiseNormSpectralModel(energy=energy)

prior = MultivariateGaussianPrior.from_(means=np.ones(10), covariance_type="diagonal")
norm.prior = prior

spectral_model = pwl * norm
```

> 

Reviewer 3: Thanks @registerrier and @adonath! I will move the use cases up and also add a similar example as you showed here!

Implementation
==============

The following implementation draft is compatible with the current API,
where the models are set via Dataset.model and the Fit is run via Fit.run(datasets).

Prior (abstract) base class:
--------------

.. code:: python 

  class Prior():
  
     def __call__(self, base):
     
        if isinstance(base, Parameter):

>             return self.evaluate(base.value)      

Reviewer 4: I guess priors should be defined in such a way, that they are unit less (i.e. a likelihood after evaluation). But maybe we should add a check for this...

        else: # model-like instance
            return self.evaluate(base.parameters.value)  
         

Future additional properties of the ``Prior`` classes: 

> - Fittable parameters

Reviewer 4: Yes, I think in general a prior should be treated much like a model. In general prior parameter can be fitted in a likelihood fit (maximum a posteriori) as well...so I think they should handle `Parameter` objects.

- Write/read from/to a yaml file, ideally also when the corresponding model is written/read (see serialisation example below)
- Prior registry system
- Prior subclass use cases (i.e. for correlated nuisance parameters with correlation matrix as a parameter)

Exemplary additional prior subclasses:
--------------
.. code:: python

  class GaussianPrior(Prior):
  
      """Gaussian Prior with mu and sigma set on a single parameter
      """
      
      def __init__(self, mu=0, sigma=1):
          self.mu = mu
          self.sigma = sigma

>       def evaluate(self, value):

Reviewer 4: Yes, I agree that the `.evaluate()` method should be static if possible. This will allows us to easier move to other tensor libraries later...The stateful call can happen in `Prior.__call__`

          return ((value - self.mu) / self.sigma) ** 2

.. code:: python

    class UniformPrior(Prior):
    
        """Uniform Prior set on a single parameter or model.
        """
        
        def __init__(self, uni=0):
            self.uni = uni

        def evaluate(self, value):
            return self.uni

.. code:: python 
  
    class CovarianceGaussianPrior(Prior):
        """Gaussian Priors on mulitple parameters with covariance matrix. Set on a model. 
        
        Parameters
        ----------
        cov : `~numpy.ndarray`
            Covariance matrix in the same order as the to be evaluated parameters.
        """
        tag = ["CovarianceGaussianPrior"]
        def __init__(self, cov):
            self.cov = cov
            
        def evaluate(self, values):
            from numpy import linalg
            return values @ linalg.inv(self.cov) @ values
        
        @property
        def prior_parameters(self):
            return [("cov", self.cov)]    
            
The prior classes are supposed to be set as ``.prior`` on:

> * Model-like classes, for instance the ``SkyModel`` class:

Reviewer 4: The reason why it should be attached to the `Model` is bookkeeping. Priors should be serialized with the model. Parameters do not know, to which model they belong. The alternative is to handle it completely independent as we do for the covariance matrix and we would rely on the order of parameters. But in this case it would still be attached to the `Model`.

  
  class SkyModel(ModelBase):
  
      _prior = None

      @property
      def prior(self):
          return self._prior

      @prior.setter
      def prior(self, value):
          self._prior = value

      def prior_stat_sum(self):
          return np.sum(self._prior(self))

* the ``Parameter`` class (for priors like ``GaussianPrior`` or ``UniformPrior``)

.. code:: python 
  
  class Parameter:
  
      _prior = None

      @property
      def prior(self):
          return self._prior

      @prior.setter
      def prior(self, value):
          self._prior = value

      def prior_stat_sum(self):
          return np.sum(self._prior(self))
    
Fit Statistic:
--------------

All priors have to be combined and added to ``Datasets.stat_sum()`` during the fitting process. 

.. code:: python 

  class Datasets():
  
      def stat_sum(self):
          """Total statistic given the current model parameters."""
          stat = self.stat_array()
          
          if self.mask is not None:
              stat = stat[self.mask.data]
              
          return np.sum(stat, dtype=np.float64) + PriorFitStatistic(weight=1).stat_sum(models = self.models)
    
    

> Where the ``PriorFitStatistic`` iterates the models and parameters, evaluates and combines the priors with the possibility to assign a weight.

Reviewer 4: The idea of the additional class is just to make it explicit, that one changes to a MAP estimate. It's like a global switch to account for the prior or not. The alternative is indeed to always evaluate priors, if present...

.. code:: python 

 class PriorFitStatistic():
       
      def __init__(self, weight=1):

>             self.weight = weight

Reviewer 4: This is the most common one, where people introduce a weight parameter (equivalent to a regularization strength...).

.. include:: ../../references.txt

.. _pig-026:

********************************
PIG 26 - Priors and FitStatistic
********************************

* Author: Noah Biederbeck
* Created: ``2023-03-13``
* Accepted:
* Status: draft
* Discussion: `#4381`_

Abstract
========

This PIG is intended to introduce priors on models that are evaluated during fitting.

Motivation
==========

> Using priors on models and on parameters is the next step in full-fledged statistical analyses.

Reviewer 5: ```suggestion
Using priors on models or on parameters is the next step in full-fledged statistical analyses.
```

> A prior can incorporate the analysers' knowledge of the topic at hand, and yield more realistically and trustworthy results.

Reviewer 5: Maybe not only the analyser... Indeed, one idea to deal with the IRFs systematics was to use the priors mechanics. So, if it is implemented like that, the knowledge comes from the people making the IRFs...

By the use, do we add as use case the inclusion of IRFs systematics into priors?

The proposed formalism also includes application of nuisance parameters and regularization.

Use cases
=========

In the past priors were a regularly requested feature or the solution to a problem.
See the following issues and PRs of Gammapy:

- Add sigma v estimator functionality for dark matter use case `#2075`_
- Background systematics as a nuisance parameter `#3955`_
- Support unfolding methods for spectral flux points `#4122`_
- Add PiecewiseNormSpatialModel `#4208`_

Implementation
==============

See `#4237`_ for an implementing PR draft.

Priors on parameters:

.. code:: python

    class GaussianPrior():
        def __init__(self, mu=0, sigma=1):
            self.mu = mu
            self.sigma = sigma

        def stat_sum(self, value):
            return ((value - self.mu) / self.sigma) ** 2

    dataset.models.parameters["x"].prior = GaussianPrior()

Priors on models:

.. code:: python

>     class L2Regularization():

Reviewer 4: Note that this case actually corresponds to the `GaussianPrior` on each of the individual parameters, where `mu=0 and sigma=1`. The most general case I could imagine would be `MultiVariateGaussianPrior` on a list of parameters. The case of `L2` regularization would be included there...but maybe we just keep it for convenience.
